### PR TITLE
Fixed menu and added photo to blog preview

### PR DESCRIPTION
**Pull request by panchtatvam**
## Issue
Blog entries on the main page currently just have title + short description. Instead, I'd like each blog link to have a big thumbnail image to the left for each blog and multiple lede lines in addition to the title. I have a list of thumbnails for about ~18 of the entries and would like a default image to be used for all other blog entries.

## Resolution
- Added `featured image` option to blog archive post at homepage
- Added `<!--more-->` command break-point for manual setting of preview text width in post content
  - Set maximum preview text limit to 500 words when manual breakpoint not set
- Added `read more` text with 3 dots to preview text to show the reader that more text can be read
- All code made responsive

## Previous merge fix
The previous [merge](https://github.com/vlad17/vlad17.github.io/pull/9) had issues which have been fixed in this code update
- Menu background shadow removed
- Inner padding of menu items reduced to match height of sibling site branding items
- Page title moved to header with required modifications
- Site subheading `dev | ml | math` kept intact. I suggest keeping it as it adds a feature to the site branding and adds a row high enough to match the height of menu and page title
- Responsiveness issues fixed

With this two tasks stand complete
1. Add menu to header
2. Add photo to blog items

Please review and merge

### CODE REVIEW

**Comment by panchtatvam** [comment]
Please view site demo at https://amazing-hoover-335b7f.netlify.app/

**Comment by panchtatvam** [comment]
## Issue 
 Instead of a "projects" page, I'd like to have a "featured" page, which contains a list of (statically selected) blog posts. This would be a subset of the main page's list of blog posts and should be maintainable via Jekyll YAML configuration.

## Resolution

- Featured page added
- Menu item created for featured page
- Icon for featured page set
- To set a post as featured, create a `is_featured_post` front matter YAML configuration in the post and set it to `true`. You can either set it `false` or just don't set this configuration at all

#### Note
The featured page doesn't have paginator like homepage has. If featured posts are more than 10, they all will show on the page instead of having a paginator navigation menu as present on the homepage. This is due to the fact that paginator.posts Jekyll loop is working only for the homepage. I was unable to make this work for featured page. If you do want this, I'll have to look at some possible workaround

Please review.

**Comment by panchtatvam** [comment]
For push [a93f91c](https://github.com/vlad17/vlad17.github.io/pull/10/commits/a93f91cc0689b7760b4b6a1188b2efc674c18213) 

## Issue
Currently, the keywords meta tag is auto-filled by Jekyll using yaml config available in each of the posts. Instead, I'd like keywords to be an explicit option in the YAML configuration on top of each of the post markdown files in the repository. This change should include adding keywords to all existing posts.

## Resolution
- Added post level custom keyword option with `meta_keywords`front matter option
- Global keywords can be set using `global_meta_keywords` option in `_config.yaml`

Please review.

**Comment by panchtatvam** [comment]
For [af7069f ](https://github.com/vlad17/vlad17.github.io/pull/10/commits/af7069f5430e8fae3940f81052d1978d0939ad18)

## Issue
Set og:image globally to a custom-made picture of a compass which includes the text "Vlad's Blog". Something made with Google Draw / powerpoint is acceptable here, I'm just looking for something when images unfurl for links to the blog on other social sites.

## Resolution
- og:image tag updated with `feature_image` tag from posts if set or `global_site_logo` tag from`_config.yaml` if not set or for homepage etc.
- transparent favicon created and deployed

Please review

**Comment by vlad17** [comment]
Thanks. All in all, looking quite good. Some fixes below.

Addressing all questions and comments in a single list:

1. Featured list is not paginated. This is OK.
2. "dev | ml | math". I see what you're saying about size. I'd like to still remove it. Could you just increase the "Vlad Feinberg" text as needed
3. The dates on the top right of each blog post entry (e.g. ,"25 Apr 2021" for the first one) have their own columns. This creates empty space and also makes the horizontal scroll look bad (see attached image). Could this date instead live in a div that contains both the blog post title (flushed left) and the date (flushed right)? This would solve both issues simultaneously.
4. I tried sharing the link on facebook, but no og:image was apparent. I'm not sure what the problem is here. Do you know what's missing? This [debugger](https://developers.facebook.com/tools/debug/?q=https%3A%2F%2Famazing-hoover-335b7f.netlify.app%2F) says content type is not set.
5. Will feature_image work for relative paths (e.g., starting with /assets)?
6. Will you update the actual requested feature_images in a separate PR? This is OK, but I want to confirm.

bad horiz scroll:
bad-horiz-scroll

no image on social:
no-img

**Comment by panchtatvam** [comment]
Responding to [requests](https://github.com/vlad17/vlad17.github.io/pull/10#issuecomment-868196339) point 2
2. Site title made bigger via [91c5eaf](https://github.com/vlad17/vlad17.github.io/pull/10/commits/91c5eafb1d3cb8da7b5279a88866557e38ad9b9a)

**Comment by panchtatvam** [comment]
Responding to [requests] point 5
5. `featured_image` (grammar fix done for the variable name) now checks if URL is internal or external and creates image link accordingly via [f94b448](https://github.com/vlad17/vlad17.github.io/pull/10/commits/f94b448307695242aba61a0fd151663d03ca067b)

**Comment by panchtatvam** [comment]
Responding to [requests](https://github.com/vlad17/vlad17.github.io/pull/10#issuecomment-868196339) point 4
4. facebook og:image was not working because site base canonical url was not set correctly. It is set to github pages URL. If set to site being deployed, it will work fine. Currently I setup up a new variable `canonical_url` just to make sure this is used only for meta tags. All other urls use `site.baseurl` jekyll inbuilt site root referencing variable. You need to set it to `vladfeinberg.com` when going live.
This is done via commit [54a7d3f](https://github.com/vlad17/vlad17.github.io/pull/10/commits/54a7d3f38bc8daa31c1ed8293cd0857858d402d4)

**Comment by panchtatvam** [comment]
I'll be adding featured_image data from the given text file in this PR to make sure that the live site keeps working with correct data after this merge. I'll update within 18 hours.

**Comment by panchtatvam** [comment]
Regarding this:
 
`The dates on the top right of each blog post entry (e.g. ,"25 Apr 2021" for the first one) have their own columns. This creates empty space and also makes the horizontal scroll look bad (see attached image). Could this date instead live in a div that contains both the blog post title (flushed left) and the date (flushed right)? This would solve both issues simultaneously.`

I'll have to reset the layout grid and edit CSS to make this work. Should I do this as this will add cost ?

File: _config.yml [hunk 1]
```diff
@@ -58,3 +59,18 @@ permalink: /:year/:month/:day/:title.html
...
+# upwork project:
+# Set custom excerpts for posts
+excerpt_separator: <!--more-->
+# vyom: 210624 Add site level meta keywords if not set
+global_meta_keywords: Vlad Feinberg, Princeton, Computer Science
+# vyom: 210624 Add og:image
+# site logo for og:image. site_url is added infront of this path. To enable external file links, {{ site_url }} needs to be removed from default.html file og:image meta tag
+global_site_logo: "/static/img/logo.jpg"
+# url for meta tags for identification of site
+canonical_url: "https://amazing-hoover-335b7f.netlify.app"
```
Comment (vlad17) [inline _config.yml]: let's please reset this back to vladfeinberg url before merge.

File: _layouts/default.html [hunk 2]
```diff
@@ -1,159 +1,192 @@
...
-                </div>
-            </div> <!-- end /.col-md-9 -->
-        </div> <!-- end /.row -->
-
-    </div> <!-- end /.container -->
-
-        <!-- Bootstrap core JavaScript
+          <div class="col-md-12 col-xs-12 footer">
+            <footer>
+              © Vladimir Feinberg 2016, 2017, 2018, 2019, 2020, 2021. Favicon by <a
```
Comment (vlad17) [inline _layouts/default.html:167]: can we make the favicon consistent with the og:image

File: _layouts/default.html [hunk 3]
```diff
@@ -1,159 +1,192 @@
...
-        <script src="{{ site.baseurl }}/static/js/material.min.js"></script>
-        <script src="{{ site.baseurl }}/static/js/main.js"></script>
-        <script src="{{ site.baseurl }}/static/js/projects.js"></script>
-    </body>
-</html>
+  <!-- Placed at the end of the document so the pages load faster -->
+
+  <script src="//ajax.googleapis.com/ajax/libs/jquery/1.11.1/jquery.min.js"></script>
+  <script src="//code.jquery.com/jquery-migrate-1.2.1.min.js"></script>
+  <script src="//maxcdn.bootstrapcdn.com/bootstrap/3.3.6/js/bootstrap.min.js"></script>
```
Comment (vlad17) [inline _layouts/default.html:183]: I can see it was this way before, but is url supposed to start with `//*` rather than `https://`?

**Review by vlad17 — APPROVED** [review]
Regarding featured image: I just confirmed it works on social. Could you please set canoncial_url back to vladfeinberg.com so that it works on merge?

Regarding new layout grid for dates: ah, it's a bit surprising that it will take a lot of work, but what you outlined seems sensible. Could you please make an estimate for how long that'd take first. If it's reasonable, then I'll give the OK and we can add this task as a separate PR.

I left a few minor 1-liner comments, do you mind please addressing them before merging?

**Comment by vlad17** [comment]
Oh, upon closer inspection, there does seem to be a larger issue: the thumbnail picture is larger than the maximum number of lede lines (see image attached). Can we make sure that thumbnails get resized to however many rem matches the lede line count?

mixed-height

**Comment by panchtatvam** [comment]
> Oh, upon closer inspection, there does seem to be a larger issue: the thumbnail picture is larger than the maximum number of lede lines (see image attached). Can we make sure that thumbnails get resized to however many rem matches the lede line count?
> 
> mixed-height

If the image mimics the height of the preview text, some images will be smaller and some larger. For now all images are of same size hence the rows are of same size. We can vertical align the text to make it vertical center ( I'll work to achieve this, if you approve ) which will make the rows all equal size and still look good.
Having said that, in case you want the images to mimic preview text height, I can do that as well. Please let me know what you would like in this case.

**Comment by panchtatvam** [comment]
> Regarding featured image: I just confirmed it works on social. Could you please set canoncial_url back to vladfeinberg.com so that it works on merge?
> 
> Regarding new layout grid for dates: ah, it's a bit surprising that it will take a lot of work, but what you outlined seems sensible. Could you please make an estimate for how long that'd take first. If it's reasonable, then I'll give the OK and we can add this task as a separate PR.
> 
> I left a few minor 1-liner comments, do you mind please addressing them before merging?

I'll take maximum 1 hour to fix the grid issue for the dates. Please let me know if I should proceed.

**Comment by panchtatvam** [comment]
I'm working on data entry for the blog images now so that nothing looks broken on the live site. Updating within 18 hours.

**Comment by vlad17** [comment]
> I'll take maximum 1 hour to fix the grid issue for the dates. Please let me know if I should proceed.

This sounds good, please proceed.

> If the image mimics the height of the preview text, some images will be smaller and some larger. For now all images are of same size hence the rows are of same size. We can vertical align the text to make it vertical center ( I'll work to achieve this, if you approve ) which will make the rows all equal size and still look good. Having said that, in case you want the images to mimic preview text height, I can do that as well. Please let me know what you would like in this case.

I see what you mean. In this case, there are indeed two changes:

1. As you suggest, let's vertically align the text to be centered.
2. The fixed image size is too large in my opinion. I think most of the blog lede lines (at least on my screen) take up ~160px on my screen, including the blog title. Perhaps the images can be closer to that? I'd be OK with anything between 160px and 200px, whatever looks best to you. Perhaps 176px (I think this comes out to 11rem), if you need a number.

**Comment by panchtatvam** [comment]
> > I'll take maximum 1 hour to fix the grid issue for the dates. Please let me know if I should proceed.
> 
> This sounds good, please proceed.
> 
> > If the image mimics the height of the preview text, some images will be smaller and some larger. For now all images are of same size hence the rows are of same size. We can vertical align the text to make it vertical center ( I'll work to achieve this, if you approve ) which will make the rows all equal size and still look good. Having said that, in case you want the images to mimic preview text height, I can do that as well. Please let me know what you would like in this case.
> 
> I see what you mean. In this case, there are indeed two changes:
> 
>     1. As you suggest, let's vertically align the text to be centered.
> 
>     2. The fixed image size is too large in my opinion. I think most of the blog lede lines (at least on my screen) take up ~160px on my screen, including the blog title. Perhaps the images can be closer to that? I'd be OK with anything between 160px and 200px, whatever looks best to you. Perhaps 176px (I think this comes out to 11rem), if you need a number.

Thank you. Working on these now.

**Comment by panchtatvam** [comment]
- Featured image data entered in posts [fca751b](https://github.com/vlad17/vlad17.github.io/pull/10/commits/fca751b0077e60f4324a9b9676b4a5d13397d0ad)
- Featured image height max capped to look better with sibling preview text [e2d293a](https://github.com/vlad17/vlad17.github.io/pull/10/commits/e2d293ad3acef03b51d3c91c4153c8785d470525)
- canonical_url set to vladfeinberg.com

Merge ready.
Working on fixing grid issue for dates in same row as heading for index page.

**Comment by vlad17** [comment]
Would you like me to merge this PR right now?

**Comment by panchtatvam** [comment]
> Would you like me to merge this PR right now?

yes please

**Comment by panchtatvam** [comment]
> Thanks. All in all, looking quite good. Some fixes below.
> 
> Addressing all questions and comments in a single list:
> 
>     1. Featured list is not paginated. This is OK.
> 
>     2. "dev | ml | math". I see what you're saying about size. I'd like to still remove it. Could you just increase the "Vlad Feinberg" text as needed
> 
>     3. The dates on the top right of each blog post entry (e.g. ,"25 Apr 2021" for the first one) have their own columns. This creates empty space and also makes the horizontal scroll look bad (see attached image). Could this date instead live in a div that contains both the blog post title (flushed left) and the date (flushed right)? This would solve both issues simultaneously.
> 
>     4. I tried sharing the link on facebook, but no og:image was apparent. I'm not sure what the problem is here. Do you know what's missing? This [debugger](https://developers.facebook.com/tools/debug/?q=https%3A%2F%2Famazing-hoover-335b7f.netlify.app%2F) says content type is not set.
> 
>     5. Will feature_image work for relative paths (e.g., starting with /assets)?
> 
>     6. Will you update the actual requested feature_images in a separate PR? This is OK, but I want to confirm.
> 
> 
> bad horiz scroll:
> bad-horiz-scroll
> 
> no image on social:
> no-img

I'm trying to fix grid issue ( point 3) but this will break so many things. So I'm doing this slowly and carefully. Shall I continue doing this ? I suggest to leave it be though. My overrides will impact so many things in the theme.

**Comment by vlad17** [comment]
Regarding the grid fix issue, it’d be helpful if you could be more specific.

Now that this base PR is merged, how about you open another one with the grid fix issue and list what you’re thinking will go wrong. A link to a live site for the PR would be helpful.
